### PR TITLE
Add multi DB support and backing up on OSS.

### PR DESCRIPTION
**Pull request by aliscott**
Will rename this repo abar-database-backup after this is merged.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -20,33 +18,61 @@ This also creates a deployment config that can later be scaled up for restoring
 
 ## Restoring from backup in AbarCloud
 
-This guide assumes you already have Postgres installed on AbarCloud. If not then see [this guide for setting it up](https://docs.abarcloud.com/additional-services/postgres.html).
+This guide assumes you already have a database installed on AbarCloud. If not then see [this guide for setting it up](https://docs.abarcloud.com/additional-services/database.html).
```
Comment (alikhajeh1) [inline README.md]: the link doesn't work, maybe just give the link to docs.abarcloud.com

File: backup.sh [hunk 2]
```diff
@@ -1,22 +1,96 @@
...
 echo "$(date): Taking backup..."
 
-# Take new backup
-pg_dump \
-   --host=${POSTGRESQL_HOST} \
-   --port=${POSTGRESQL_PORT} \
-   --username ${POSTGRESQL_USER} \
-   --dbname=${POSTGRESQL_DATABASE} \
-   --format=custom > backups/${POSTGRESQL_DATABASE}-$(date +"%Y%m%d%H%M%S").dump
+# Loop through each entr in the config file
```
Comment (alikhajeh1) [inline backup.sh]: entry

File: backup.sh [hunk 3]
```diff
@@ -1,22 +1,96 @@
...
+  host=$(echo $backup_config | jq -r ".[$i].host")
+  port=$(echo $backup_config | jq -r ".[$i].port")
+  user=$(echo $backup_config | jq -r ".[$i].user")
+  password=$(echo $backup_config | jq -r ".[$i].password")
+  database=$(echo $backup_config | jq -r ".[$i].database")
+  prefix=$(echo $backup_config | jq -r ".[$i].prefix")
+  keep=$(echo $backup_config | jq -r ".[$i].backups_to_keep")
+
+  # Generate a backup filename
+  backup_filename=${prefix}$(date +"%Y%m%d%H%M%S").dump
```
Comment (alikhajeh1) [inline backup.sh]: should we chuck UTC to the end so it's explicit what TZ it's using?

File: backup.sh [hunk 4]
```diff
@@ -1,22 +1,96 @@
...
+
+  # Continue to next config if there's an error
+  if [[ -n "$error" ]]; then
+    echo "$(date): ERROR: Error taking backup of $type database $database: $error" >&2
+    continue
+  fi
+
+  # Upload to OSS
+  echo "$(date): Uploading backup for $type database $database to object storage..."
+  ossutil cp backups/${backup_filename} oss://$OSS_PATH/$backup_filename
```
Comment (alikhajeh1) [inline backup.sh]: Have we tried uploading a single large file (e.g. 1GB+) to see if uses multipart upload by default?

File: database-backup.yml [hunk 5]
```diff
@@ -47,46 +53,39 @@ objects:
...
-                value: "${BACKUPS_TO_KEEP}"
-              image: registry.abarcloud.com/library/postgresql-backup:latest
+              - name: OSS_PATH
+                value: "${OSS_PATH}"
+              - name: BACKUP_CONFIG
+                value: "/etc/backup-config/config.json"
+              image: registry.abarcloud.com/library/database-backup:latest
               resources:
                 limits:
                   memory: 200Mi
```
Comment (alikhajeh1) [inline database-backup.yml]: Does this work ok with large dumps? would 100Mi be enough?

File: README.md [hunk 6]
```diff
@@ -3,14 +3,12 @@
 You can create the cronjob, secret and volume required in your AbarCloud project by processing the template provided:
```
Comment (alikhajeh1) [inline README.md]: Should this mention that users first need to clone this repo, edit the json file with their configs then run the command?

We could also mention how they could update the configs (edit the secret then click on Deploy).

File: backup.sh [hunk 7]
```diff
@@ -1,22 +1,96 @@
...
+  mysqldump \
+     --host=${host} \
+     --port=${port} \
+     --user=${user} \
+     --compress \
+     ${database} > backups/${backup_filename}
+}
+
+function backup_mongodb() {
+  mongodump \
```
Comment (alikhajeh1) [inline backup.sh]: https://docs.mongodb.com/manual/reference/program/mongodump/ 
> However, the use of mongodump and mongorestore as a backup strategy can be problematic for sharded clusters and replica sets.

https://stackoverflow.com/a/22197164 says the user can specify a replicaSet connection string. 

Should we add a link to https://docs.mongodb.com/manual/reference/program/mongodump/ in the README and suggest that MongoDB users think carefully about how they want to back up? e.g. they might want to backup one of the nodes in the replicaSet or use this replicaSet string option? We can maybe just add this note and see how users use this then update it with more details.

File: database-backup.yml (hunk 5 above)
Comment (aliscott) [reply database-backup.yml]: 200Mi seemed correct for smaller dumps. For larger dumps the memory needs to be increased. Will add a note to the README.

File: backup.sh (hunk 4 above)
Comment (aliscott) [reply backup.sh]: ossutil uses resumable uploads for anything over 100Mi by default.

**Pull request merged**